ping-payload-extensions/ping-custom-payload-extensions.md

```diff
@@ -0,0 +1,67 @@
+# Ping Custom Payload Extensions
+
+## The problem
+
+Over time we will need to change what is sent over ping/pong `custom_payload`'s to adjust for what is needed in the future.
+Currently the only way to update a `custom_payload` is through breaking changes which break protocol compatible between Portal implementations.
+As we get users, deploying breaking changes to mainnet will no longer be an option as we are aiming for a 100% uptime.
+
+## The Solution
+
+Ping Custom Payload Extensions. A versioned type prefixed format where we can upgrade endpoints on a new version well giving a window for all Portal implementations to update before deprecating the older version. This will allow clients to implement new functionality without breaking compatibility with the standard specification.
+
+# Type's
+
+There are 4_294_967_296 unique type ids.
+
+Types 0-10_000 and 4_294_957_295-4_294_967_296 are reserved for for future upgrades.
+
+The rest are first come first serve, but they should still be defined in this repo to avoid overlaps.
+
+
+## Requirements
+
+All payloads used in the Ping `custom_payload` MUST follow the `Ping Custom Payload Extensions` format.
+
+## Custom Payload Extensions Format
+
+- **type**: what payload type is it
+- **verison**: what version of the type it is
+- **payload**: a ssz container which contents are specified by the type and version field
+
+
+```python
+CustomPayloadExtensionsFormat = Container(
+  type: Bytes4,
+  version: Bytes4,
+  payload: Container(inner payload is defined by type and version)
+)
```

Comment:
The way this currently is defined, it means that one has to write some custom code to only decode the encoded type and version, at which point it can then know how to decode the other `Container`.

Unless you meant to have the `payload` to be a `ByteList` which then could be decoded in a second step based on the type/version information after decoding the `CustomPayloadExtensionsFormat` `Container`.  That would be a possible solution for this.
Or as `Ping.payload` is already a `ByteList`, another way might be to just prefix it with those bytes that indicated the type & version. Either way would work.